src/SALib/analyze/morris.py

```diff
@@ -73,214 +73,414 @@ def analyze(problem, X, Y,
+    -------
+    sigma: np.ndarray
+        Sigma values for the groups.
+    """
+    sigma_agg = _compute_grouped_metric(ungrouped_sigma, groups)
+
+    sigma = np.zeros(groups.shape[1], dtype=np.float)
+    np.copyto(sigma, sigma_agg, where=groups.sum(axis=0) == 1)
+    np.copyto(sigma, np.NAN, where=groups.sum(axis=0) != 1)
 
     return sigma
 
 
-def compute_grouped_metric(ungrouped_metric, group_matrix):
-    '''
-    Computes the mean value for the groups of parameter values in the
-    argument ungrouped_metric
-    '''
+def _compute_grouped_metric(ungrouped_metric: np.ndarray,
+                            groups: np.ndarray) -> np.ndarray:
+    """ Computes the mean value for the groups of parameter values.
 
-    group_matrix = np.array(group_matrix, dtype=np.bool)
+    Parameters
+    ----------
+    ungrouped_metric: np.ndarray
+        Metric calculated without considering the groups
+    groups: np.ndarray
+        Array defining the distribution of groups
 
-    mu_star_masked = np.ma.masked_array(ungrouped_metric * group_matrix.T,
-                                        mask=(group_matrix ^ 1).T)
-    mean_of_mu_star = np.ma.mean(mu_star_masked, axis=1)
+    Returns
+    -------
+    mean_of_mu_star: np.ndarray
+         Mean value for the groups of parameter values
+    """
 
-    return mean_of_mu_star
+    groups = np.array(groups, dtype=np.bool)
 
+    mu_star_masked = np.ma.masked_array(ungrouped_metric * groups.T,
+                                        mask=(groups ^ 1).T)
+    mean_of_mu_star = np.ma.mean(mu_star_masked, axis=1)
 
-def get_increased_values(op_vec, up, lo):
+    return mean_of_mu_star
 
-    up = np.pad(up, ((0, 0), (1, 0), (0, 0)), 'constant')
-    lo = np.pad(lo, ((0, 0), (0, 1), (0, 0)), 'constant')
 
-    res = np.einsum('ik,ikj->ij', op_vec, up + lo)
+def _reorganize_output_matrix(output_array: np.ndarray,
+                              value_increased: np.ndarray,
+                              value_decreased: np.ndarray,
+                              direction: str) -> np.ndarray:
+    """Reorganize the output matrix.
 
-    return res.T
+    This method reorganizes the output matrix in a way that allows the
+    elementary effects to be computed as a simple subtraction between two
+    arrays. It repositions the outputs in the output matrix according to the
+    order they changed during the formation of the trajectories.
 
+    Arguments
+    ----------
+    output_array: np.ndarray
+        Matrix of model output values
+    value_increased: np.ndarray
+        Input variables that had their values increased when forming the
+        trajectories matrix
+    value_decreased: np.ndarray
+        Input variables that had their values decreased when forming the
+        trajectories matrix
+    direction: str
+        Direction to consider (values that increased or decreased)
+    Returns
+    -------
 
-def get_decreased_values(op_vec, up, lo):
+    """
+    if direction == "increased":
```

Comment:
Might be worth using an `enum` or `bool` here rather than a `str` as a string comparison is relatively expensive.

Reply:
Done! I've used `bool`.